### ISSUE TITLE
Pure Data

### ISSUE THREAD

**Issue by mxmilkb** [github/linguist#4508]
## Preliminary Steps

yes/done

## Problem Description
PD files aren't detected and listed

### URL of the affected repository:
Examples:

https://github.com/MikeMorenoAudio/EP-MK1

https://github.com/danomatika/BangYourHead

### Last modified on:
15th Feb

29th Jan

### Expected language:
Pure Data

### Detected language:
Nothing

**Comment by pchaigno** [comment]
Looks like these `.pd` files [are detected as Pure Data](https://github.com/MikeMorenoAudio/EP-MK1/search?l=Pure+Data). They don't show up in the language bar because GitHub doesn't display "data languages" by default. You can use overrides to change that.

**Comment by mxmilkb** [comment]
A-ha, thank you for the clarification.

I've just checked through my GH settings, but I'm not sure what you mean by overrides? Or do you mean if I run/host linguist?

I guess this repo isn't for making GH policy suggestions, but might you be able to advise if GH notes the definition of a "data language" somewhere? Is this somehow related to the notion of dataflow languages?

**Comment by Alhadis** [comment]
@mxmilkb Your questions are already answered by the steps you neglected to follow in the issue template:

> ## Preliminary Steps
> yes/done


Here they are again for your reference:

> Please confirm you have...
> - [ ] reviewed [How Linguist Works](https://github.com/github/linguist#how-linguist-works),
> - [ ] reviewed the [Troubleshooting](https://github.com/github/linguist#troubleshooting) docs,
> - [ ] considered implementing an [override](https://github.com/github/linguist#overrides),
> - [ ] verified an issue has not already been logged for your issue ([linguist issues](https://github.com/issues?utf8=%E2%9C%93&q=is%3Aissue+repo%3Agithub/linguist)).
>
> *Please review these preliminary steps before logging your issue. __You may find the information referenced may answer or explain the behaviour you are seeing.__ It'll help us to know you've reviewed this information.*

**Comment by mxmilkb** [comment]
Thank you for the prompt reply, but possibly you fail to see the ambiguity here?

GitHub the service could have per-user hooks into linguist overrides, but I take it that it doesn't.

I've just double checked, but none of the four steps say why .pd files are defined as a "data" language".

**Comment by Alhadis** [comment]
The `data` type is a bit of a catchall, but generally used for languages which aren't written by hand (or those which are, but aren't programming, document markup, or lightweight markup languages; e.g., most configuration files).

[Pure Data's homepage](https://puredata.info/) states (emphasis mine):

> Pd enables musicians, visual artists, performers, researchers, and developers to create software graphically **_without writing lines of code._**

For program-generated formats like this, the `data` type is an apt description. If you want it considered as part of a project's statistics, we recommend using the `linguist-detectable` override.

**Comment by mxmilkb** [comment]
Other visual dataflow programming languages like Max (created by the same person as Pure Data) patches are [displayed](https://github.com/Cycling74/n4m-examples), also are [LabVIEW](https://github.com/picotech/picosdk-ni-labview-examples). Pure Data is certainly [Turing complete](http://abstrakt.vade.info/?p=92) if that matters. I would contend that `data` is not an apt description (that the definition of `data` is 'undefined' is problematic and worthy of it's own issue IMHO).

**Comment by Alhadis** [comment]
> Other visual dataflow programming languages like Max (created by the same person as Pure Data) patches are displayed, also are LabVIEW.

That might need investigation then, especially if the formats are simply [JSON](https://github.com/Cycling74/n4m-examples/blob/2b0df03891240e55b2201279f6cdf5d49bf1bf42/express/express.maxproj) and [XML](https://github.com/JKISoftware/JKI-State-Machine-Objects/blob/f2f3660cb86d0283772416fe1704513aa1f808a3/tests/JKI%20SMO%20Example.lvproj). It's possible these should have been added as file-extensions of JSON and XML, respectively (unless there's an obvious syntactic difference).

> I would contend that `data` is not an apt description (that the definition of `data` is 'undefined' is problematic and worthy of it's own issue IMHO).

No, the issue here is that Linguist and GitHub have no concept of a "dataflow" or "visual" programming language. We've experienced a similar issue to this one when Eagle/KiCad projects [were redefined](https://github.com/github/linguist/pull/3743) as `data` instead of `markup`. While the change was justified, it [was also disruptive](https://github.com/github/linguist/issues/3784) for KiCad/Eagle users, whose projects were suddenly lacking any classification whatsoever. Like Pure Data, these weren't files that were created by humans, but through a graphical interface.

The solution to the KiCad crisis was the introduction of the `linguist-detectable` override, which is our best possible solution to machine-generated formats that users expect to be part of a project's classification.

**Comment by Alhadis** [comment]
@lildude @pchaigno Any thoughts on what to do with [Max](https://github.com/github/linguist/blob/9116c90fcbb82ac03b4b33c58cfbde1fcf745e99/lib/linguist/languages.yml#L2968) and [LabVIEW](https://github.com/github/linguist/blob/9116c90fcbb82ac03b4b33c58cfbde1fcf745e99/lib/linguist/languages.yml#L2566-L2574)? It's clear they shouldn't be `programming`, and or even registered as languages (assuming there's no incompatibilities with JSON/XML, respectively).

**Comment by mxmilkb** [comment]
Massive apologies to Max and LabVIEW users reading this later if it comes to pass that this weird subjective classification of visual programming language source files affects your user experience of GitHub. Please don't hate me, for I will not have been the one making the change.

I can see the rational for KiCad being classed as `data`, but I don't think any of the points to raise on that issue relate to this issue.

**Comment by mxmilkb** [comment]
Ah, this is interesting - https://github.com/proteusvacuum/KlattSynth - does list Pure Data files in the language bar..

**Comment by lildude** [comment]
@mxmilkb that'll be because that repo hasn't been touched in over 3 years. In the time since, Pure Data was switched to being classified as data by @Alhadis in https://github.com/github/linguist/pull/3751. Repositories are not reanalysed when Linguist is updated, only when changes are pushed to the repo, hence that repo is still showing the analysis from before @Alhadis's changes were merged and rolled out.

**Comment by mxmilkb** [comment]
I would now summarise that the issues at hand are:

- The conception of a 'data' format is an undefined catch-all 
- Visual programming language files are incorrectly not classed as `programming`

> The data type is a bit of a catchall, but generally used for languages which aren't written by hand (or those which are, but aren't programming, document markup, or lightweight markup languages; e.g., most configuration files).

"languages which aren't written by hand"

Can I request a rational for that distinction?

Yes, you edit them with an IDE of sorts, but it's still a symbolic programming language at the IDE level. A related example could be Smalltalk, where the concept of an IDE is generally integral to the use of the language.

I'm also confused as to why the feedback on the PR wasn't replied to.

**Comment by Alhadis** [comment]
> Yes, you edit them with an IDE of sorts, but it's still a symbolic programming language at the IDE level.

And where exactly would you draw the line between "visual programming" and "flowchart / mind-mapping software"? If we classify PD as "programming", how do we justify to users why Xcode storyboards and Android layouts (both XML) are "data", despite subjective overlap? See #2818 and #3125 for prior discussion about the latter.

GitHub's userbase spans 28 million people with projects spanning every fathomable interest and background. We can't hope to please everybody, so we give them the option of overriding aspects of Linguist they disagree with, or simply don't fit their projects in question.

**Comment by mxmilkb** [comment]
Well, it has inputs, outputs, performs computations to process data between those inputs and outputs. As mentioned before, it's uses the dataflow paradigm, similar to the functional reactive paradigm. It has control flow mechanisms, data types and data structures. It has a standard library as well as external modules to add more functions. Aside from the history of use of the word "programming" already being associated with both Pure Data and the general concept of visual programming languages, Pure Data can do "programming" things, from [hello world](https://www.youtube.com/watch?v=rtgGol-I4gA) to [advanced DSP](https://www.youtube.com/watch?v=l18q2F5FXq8), making [sequenced/synthesised](https://www.youtube.com/watch?v=9Nz0bxwoqQE) and [generative](https://www.youtube.com/watch?v=0xr4aL1C24E) music, for [livecoding](https://www.youtube.com/watch?v=jADuLtsFqkk), as a [UI toolkit](https://www.youtube.com/watch?v=Kp9PdmPSBik), for processing [shaders](https://www.youtube.com/watch?v=58QKFQy9ua8) or controlling [lighting rigs](https://www.youtube.com/watch?v=3Q4hLA7YM7c).

**Comment by Alhadis** [comment]
> making sequenced/synthesised and generative music, for livecoding, as a UI toolkit, for processing shaders or controlling lighting rigs

Yeah, that's where the problems begin. There are *many* programs which can do that... Xquartz, AudioMulch, Substance, countless procedural texture generators, and even compositing/post-editing suites like Flame and After Effects have railroad/"flow"-like diagrams to programmatically control input/output. These are all programs I've worked with, too, and I know how technical and complex they can become. How are they any different? Because of how the project's data is represented on disk?

One of Linguist's [advertised responsibilities](https://github.com/github/linguist#linguist) is identification of machine-generated formats:

> This library is used on GitHub.com to **detect blob languages**, ignore binary or vendored files, **suppress generated files** in diffs, and generate language breakdown graphs.

"Machine-generated" here basically meaning "something you didn't write by hand using a text-editor", which leaves no room for ambiguity. Most users consider this desirable, and those who don't have the overrides facility to address projects on a case-by-case basis. The system is simple, but it works.

Now, the credibility of a "visual programming language" is irrelevant here. The key factor is that Pure Data is a program-generated format not designed to be written by hand. This is *precisely* what Linguist classifies as a "generated file", and if we don't enforce consistency, everything leads to a slippery slope of vague excuses to the next user who tries "lawyering" their favourite program's file-format because it was authored using a similar-looking interface.

**Comment by mxmilkb** [comment]
Touching on the key factor first;

> suppress generated files

> The key factor is that Pure Data is a program-generated format not designed to be written by hand. This is precisely what Linguist classifies as a "generated file"

Where does Linguist define that precisely to be a "generated file"? Does that not [mean](https://github.com/github/linguist/blob/0dc5bf140eb89457ddff64afeddb3ae4ff2a29a2/test/test_generated.rb#L39) the like of build artefacts generated by package managers, frameworks, preprocessing and compilation?

A .pd file is something one edits, saves, and then edits again, and can run in various ways (see below).

> detect blob languages

Why have you highlighted this? In the context of linguist, does 'blob' not refer to an [unclassified](https://github.com/github/linguist/blob/972669ea5bf9f88b9560d22333dba26977c9df04/lib/linguist/heuristics.rb#L8) [file](https://github.com/github/linguist/blob/a5df9a00ab7c9828bd7038bb9f9bd5e56d325dc9/lib/linguist/strategy/modeline.rb#L109)? Or actually the traditional definition of blob, i.e., binary or bytecode blobs (as Assembly and LLVM are classed as languages)?

Assembly and LLVM bytecode are, very much for the most part, not written by hand.

> Yeah, that's where the problems begin

I think a problem might better be framed as something that can be actioned against, i.e., not having a stronger definitions of the categories involved, rather than framed against the existence of various forms of software that are fairly generally accepted to be on a different level to visual programming languages.

> Xquartz

Which is X for Mac? A display server, so you write code in a language that uses X protocols, so it's appropriate for that code to be classed as the language it's written in, no?

> AudioMulch, Substance, countless procedural texture generators, and even compositing/post-editing suites like Flame and After Effects

AudioMulch I know of but have not used, though [here's a thread](https://forum.ableton.com/viewtopic.php?t=117574) with some folk experienced in both saying that the like of Max and Pd are really on a different level to AudioMulch.

In general, visual programming languages are on a different and lower level to the like of texture generation or video compositing software.

> program's file-format 

There are [multiple](https://puredata.info/downloads) [implementations](https://github.com/jonwwilkes/pd-nw) [of](https://git.purrdata.net/jwilkes/purr-data) [Pure](https://github.com/libpd/libpd) [Data](https://github.com/sebpiq/WebPd) [available](https://github.com/pierreguillot/Camomile) [to](https://github.com/x37v/pdlv2) [use](https://github.com/jacobstern/ZenGarden) [patches](http://droidparty.net/) [with](https://github.com/sharebrained/pd_compiler), so it is not like a one program format.

**Comment by Alhadis** [comment]
> Does that not mean the like of build artefacts generated by package managers, frameworks, preprocessing and compilation?

Yes, those files would normally be included in project statistics, so we purposefully exclude them. For a language/format that's *entirely* generated, it makes no sense it register it as "programming" and then mark each and every single file as "generated". Therefore, we classify it as `data` which excludes it from classification.

>Which is X for Mac?

I meant \*Quartz composer, sorry. Currently working with X at the moment… guess that was an easy brainfart to make…

> A .pd file is something one edits, saves, and then edits again, and can run in various ways (see below).

Yes, but the contents of the file **weren't written by hand**, and aren't supposed to be. That's what we define as "generated". I've explained the solution to you (repeatedly), and fine, you disagree. If you continue to argue I'm going to have to lock this thread.

> Assembly and LLVM bytecode are, very much for the most part, not written by hand.

GitHub hosts hundreds (if not *thousands*) of historic codebases from the pre-Unix/C era, [some of which](https://github.com/chrislgarry/Apollo-11) are quite famous. Moreover, "Assembly" covers many languages, both modern and historic, auto-generated and human-written, so it's not fair to make the assumption it's "for the most part not written by hand".

> In general, visual programming languages are on a different and lower level to the like of texture generation or video compositing software.

Given that After Effects embeds an entire JavaScript engine to run embedded keyframe programs, I'm inclined to disagree with you. In fact, an [interactive animation](https://github.com/Alhadis/REST-GreenWalk) I rendered was keyframed for another language using After Effects' JS-generated output. 

I disgress.
